### PR TITLE
[build] Turn on function level linking for taichi_c_api

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: https://github.com/taichi-dev/taichi/issues/6793

### CODE REVIEW

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

File: cmake/TaichiCAPI.cmake [hunk 1]
```diff
@@ -19,9 +19,9 @@ endif()
 
 endfunction()
 
-
 set(TAICHI_C_API_NAME taichi_c_api)
 
+set(TAICHI_C_API_NAME taichi_c_api)
```
Comment (PENGUINLIONG) [inline cmake/TaichiCAPI.cmake:24]: Is it duplicated?

File: CMakeLists.txt [hunk 2]
```diff
@@ -64,6 +66,8 @@ option(USE_LLD "Use lld (from llvm) linker" OFF)
 option(USE_MOLD "Use mold (A Modern Linker)" OFF)
 option(TI_WITH_BACKTRACE "Use backward-cpp to print out C++ stack trace upon failure" OFF)
 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ffunction-sections -fdata-sections")
```
Comment (PENGUINLIONG) [inline CMakeLists.txt:69]: Does it work with msvc?

Comment (jim19930609) [reply CMakeLists.txt:69]: hmmm, let me limit it to Linux and Mac only

**Review by PENGUINLIONG — APPROVED** [review]
Thank you! 🤗

**Pull request merged**